### PR TITLE
Fix spatial frequency for non-square gratings

### PR DESCRIPTION
**Pull request by znamensk**
Proposed fix for issue #7.

`Grating.vert` now simply returns `vt` as `texCoord`. `Grating.frag` has separate `frequency_x` and `frequency_y` uniforms and grating orientation is set by adjusting them in the workflow. 

The alternative would be to adjust the way SF and Angle are calculated in the workflow before being passed to the shader, but I think separating `frequency_x` and `frequency_y` is more readable and easier to maintain.

### CODE REVIEW

**Comment by glopesdev** [comment]
Finally got a chance to review this, overall it looks great. I have only a couple of optimizations such as implementing `frequency` using a `vec2` uniform instead of two separate uniforms. This also allows refactoring the `SetFrequency` group to use only one `Zip` operator rather than two, similar to the original implementation.

However, I noticed one downside of this approach. Using this solution, if you draw a circular grating and continuously animate the `Angle` property, the gratings will not rotate around the center of the gratings, but rather seem to show an offset pivot. I'm still trying to wrap my head around why this happens and how to fix it.

I think this was the main reason why I switched to using a texCoord rotation, to avoid having to think about grating orientation in the fragment shader. Will update when I make progress.

**Comment by znamensk** [comment]
Hmm... Good point. In the amended version gratings currently rotate around (0,0) in UV coordinates. I need to test it but I think simply subtracting 0.5 from `texCoord` along both dimensions in the vertex shader should make them rotate around the centre?

**Review by glopesdev — CHANGES_REQUESTED** [review]
Centring the texture coordinates does the trick, so in that case below are the only changes I would do before merging.


Replace the two frequency uniforms with a `vec2` field:
https://github.com/bonvision/BonVision/blob/23b933c90ef995a6d40003e5b8208effece815ab/BonVision/Shaders/Gratings.frag#L6-L7

```glsl
uniform vec2 frequency;
```

The value calculation then has to be updated to take into account centring as you suggested:
https://github.com/bonvision/BonVision/blob/23b933c90ef995a6d40003e5b8208effece815ab/BonVision/Shaders/Gratings.frag#L17

```glsl
  float value = (texCoord.x - 0.5) * frequency.x + (texCoord.y - 0.5) * frequency.y + phase;
```

Finally, having a single uniform allows simplifying the `SetFrequency` group to use only a single `Zip` operator, like so:
image

Where the `ExpressionTransform` operator reads:
```c#
new(
single(Item1 * Math.Cos(Item4) * Item3) as Fx,
single(Item2 * Math.Sin(Item4) * Item3) as Fy)
```

**Comment by glopesdev** [comment]
@znamensk I was curious to try the suggested edits feature of GitHub. Can you check if your branch is set to allow edits from maintainers? If it is, I should be able to suggest the changes myself by directly adding commits to the forked branch.

On the GitHub doc pages there is more info about [allowing changes to a pull request branch created from a fork]( https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).

**Comment by znamensk** [comment]
I somehow don't see the **Allow edits from maintainers** checkbox from my end... From what I understood from Github docs, it should be on this pull request - I made the pull request directly to this upstream repo.

**Comment by glopesdev** [comment]
Strange... in that case, maybe I can merge the pull request as it is and then take it from there, if that works for you?

**Comment by znamensk** [comment]
Sure, of course.

**Pull request merged**
